src/interfaces/asymmetric_crypto.rs

```diff
@@ -0,0 +1,382 @@
+#![allow(dead_code)]
+use std::{convert::TryFrom, sync::Mutex};
+
+use cosmian_crypto_base::{
+    asymmetric::{AsymmetricCrypto, KeyPair},
+    entropy::CsRng,
+    hybrid_crypto::{BytesScanner, Header, Metadata},
+    symmetric_crypto::{aes_256_gcm_pure::Aes256GcmCrypto, Key, Nonce, SymmetricCrypto},
+};
+use rand_core::RngCore;
+
+use crate::{
+    core::{gpsw::AbeScheme, Engine},
+    interfaces::policy::{AccessPolicy, Attribute, Policy},
+};
+
+#[derive(Clone, PartialEq)]
+pub struct AbeMasterKeys<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    pub master_private_key: S::MasterPrivateKey,
+    pub public_key: S::MasterPublicKey,
+    pub master_public_delegation_key: S::MasterPublicDelegationKey,
+}
+
+#[derive(Clone, PartialEq)]
+pub struct AbeKeyPair<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    pub public_key: S::MasterPublicKey,
+    pub private_key: S::UserDecryptionKey,
+}
+
+impl<S> KeyPair for AbeKeyPair<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    type PrivateKey = S::UserDecryptionKey;
+    type PublicKey = S::MasterPublicKey;
+
+    fn public_key(&self) -> &Self::PublicKey {
+        &self.public_key
+    }
+
+    fn private_key(&self) -> &Self::PrivateKey {
+        &self.private_key
+    }
+}
+
+pub struct PrivateKeyGenerationParameters<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    pub master_private_key: S::MasterPrivateKey,
+    pub policy: Policy,
+    pub access_policy: AccessPolicy,
+}
+
+pub struct KeyPairGenerationParameters<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    pub master_private_key: S::MasterPrivateKey,
+    pub policy: Policy,
+    pub master_public_key: S::MasterPublicKey,
+    pub access_policy: AccessPolicy,
+}
+
+pub struct EncryptionParameters {
+    pub policy: Policy,
+    pub policy_attributes: Vec<Attribute>,
+}
+
+pub struct AbeCrypto<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    rng: Mutex<CsRng>,
+    engine: Engine<S>,
+}
+
+impl<S> Default for AbeCrypto<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<S> AbeCrypto<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    // fn ensure_engine(&self) -> anyhow::Result<&Engine<S>> {
+    //     self.engine.as_ref().ok_or_else(|| {
+    //         anyhow::anyhow!(
+    //             "The Policy must first be set on ABE scheme using set_scheme_parameters"
+    //         )
+    //     })
+    // }
+
+    /// Generate an ABE master key pair for the Policy, returning a triple
+    /// (private key, public key, public delegation key)
+    fn generate_master_keys(&self, policy: &Policy) -> anyhow::Result<AbeMasterKeys<S>> {
```

Comment:
```suggestion
    [cfg(test)]
    fn generate_master_keys(&self, policy: &Policy) -> anyhow::Result<AbeMasterKeys<S>> {
```

Reply:
Not sure whether this should or should not be publibly available. Will limit visibility to test as suggested.